### PR TITLE
Array conditionals

### PR DESCRIPTION
**Pull request by agonza40**
### This project uses [semver](semver.org), please check the scope of this pr:

- [ ] #none# - documentation fixes and/or test additions
- [ ] #patch# - backwards-compatible bug fix
- [x] #minor# - adding functionality in a backwards-compatible manner
- [ ] #major# - incompatible API change

# CHANGELOG

* **Added** support for array conditions in both models and views.

* **Added** view model extension support. This means you can do something like the following:

  *Model*

  ```json
  {
    "properties": {
      "lang": {
        "type": "string"
      }
    },
    "type": "object"
  }
  ```

  *View*
  ```json
  {
    "cells": [
      {
        "children": [
          {
            "id": "langGroup",
            "model": {
              "enum": [
                "Server",
                "Web"
              ],
              "type": "string"
            },
            "internal": true
          },
          {
            "conditions": [
              {
                "if": [
                  {
                    "_internal.langGroup": {"equals": "Web"}
                  }
                ]
              }
            ],
            "model": "lang",
            "renderer": {
              "data": [
                {"label": "CSS", "value": "css"},
                {"label": "HTML", "value": "html"},
                {"label": "JavaScript", "value": "js"}
              ],
              "name": "select"
            }
          },
          {
            "conditions": [
              {
                "if": [
                  {
                    "_internal.langGroup": {"equals": "Server"}
                  }
                ]
              }
            ],
            "model": "lang",
            "renderer": {
              "data": [
                {"label": "Express", "value": "express"},
                {"label": "Python", "value": "py"},
                {"label": "Ruby", "value": "rb"}
              ],
              "name": "select"
            }
          }
        ]
      }
    ],
    "type": "form",
    "version": "2.0"
  }
  ```

### CODE REVIEW

File: addon/components/array-container.js [hunk 1]
```diff
@@ -12,6 +12,24 @@ import layout from 'ember-frost-bunsen/templates/components/frost-bunsen-array-c
 
 const {keys} = Object
 
+function itemIndex (arrayId, itemId) {
+  return itemId.slice(arrayId.length + 1).split('.')[0]
+}
+function clearSubObject (object, path, pathIndex) {
```
Comment (sandersky) [inline addon/components/array-container.js]: JSDoc

File: addon/components/array-container.js [hunk 2]
```diff
@@ -131,29 +159,71 @@ export default Component.extend(HookMixin, PropTypeMixin, {
...
     }
 
     const items = get(value || {}, bunsenId) || []
-
+    const getModel = Array.isArray(bunsenModel.items)
+      ? (item, index) => bunsenModel.items[index] || bunsenModel.additionalItems
+      : () => bunsenModel.items
+    const getCellConfig = Array.isArray(cellConfig)
+      ? (item, index) => cellConfig[index] || cellConfig[cellConfig.length - 1] || {}
+      : () => cellConfig || {}
```
Comment (sandersky) [inline addon/components/array-container.js]: Not a huge fan of this multiline ternary going on. I'd almost prefer something like:

```js
let getModel, getCellConfig

if (Array.isArray(bunsenModel.items) {
  getModel = (item, index) => bunsenModel.items[index] || bunsenModel.additionalItems
} else {
  getModel = () => bunsenModel.items
}

if (Array.isArray(cellConfig) {
  getCellConfig = (item, index) => cellConfig[index] || cellConfig[cellConfig.length - 1] || {}
} else {
  getCellConfig = () => cellConfig || {}
}
```

@cstolli do you have an opinion on this?

Comment (agonza40) [inline addon/components/array-container.js]: Yeah this really makes me wish we had conditional expressions like they do in some other languages (*cough* Elm *cough*)

**Comment by sandersky** [comment]
:+1:

**Pull request merged**